Modules/mmapmodule.c

```diff
@@ -503,51 +505,96 @@ mmap_resize_method(mmap_object *self,
-        /* Create another mapping object and remap the file view */
+        /* if it's not the paging file, unmap the view and resize the file */
+        if (self->file_handle != INVALID_HANDLE_VALUE) {
+            if (!UnmapViewOfFile(self->data)) {
+                return PyErr_SetFromWindowsErr(GetLastError());
+            };
+            self->data = NULL;
+            /* resize the file */
+            if (!SetFilePointerEx(self->file_handle, max_size, NULL,
+                FILE_BEGIN) ||
+                !SetEndOfFile(self->file_handle)) {
+                /* resizing failed. try to remap the file */
+                error = GetLastError();
+                new_size = max_size.QuadPart = self->size;
+            }
+        }
+        /* create a new file mapping and map a new view */
+        /* FIXME: call CreateFileMappingW with wchar_t tagname */
         self->map_handle = CreateFileMapping(
             self->file_handle,
             NULL,
             PAGE_READWRITE,
-            0,
-            0,
+            max_size.HighPart,
+            max_size.LowPart,
             self->tagname);
-        if (self->map_handle != NULL) {
-            self->data = (char *) MapViewOfFile(self->map_handle,
-                                                FILE_MAP_WRITE,
-                                                off_hi,
-                                                off_lo,
-                                                new_size);
+        if (self->map_handle != NULL &&
+            GetLastError() != ERROR_ALREADY_EXISTS) {
+            self->data = MapViewOfFile(self->map_handle,
+                FILE_MAP_WRITE,
+                offset.HighPart,
+                offset.LowPart,
+                new_size);
             if (self->data != NULL) {
+                /* copy the old view if using the paging file */
+                if (self->file_handle == INVALID_HANDLE_VALUE) {
+                    memcpy(
+                        self->data, old_data,
+                        self->size < new_size ? self->size : new_size
+                    );
+                }
                 self->size = new_size;
-                Py_RETURN_NONE;
-            } else {
-                dwErrCode = GetLastError();
+            }
+            else {
+                error = GetLastError();
                 CloseHandle(self->map_handle);
                 self->map_handle = NULL;
             }
-        } else {
-            dwErrCode = GetLastError();
         }
-        PyErr_SetFromWindowsErr(dwErrCode);
-        return NULL;
+        else {
+            error = GetLastError();
+        }
+        /* unmap the old view if using the paging file
+        Don't do this if we've encountered ERROR_ALREADY_EXISTS as this means
+        that creating a new named file mapping failed and the mapping handle
+        points to an existing mapping of the same name
+        */
+        if ((GetLastError() != ERROR_ALREADY_EXISTS) &&
```

Comment:
Should this compare against `error` rather than calling `GetLastError()` again?

In fact, it definitely should, because `GetLastError()` _is not_ guaranteed to return 0 when the last call was successful - it only updates on errors.

```suggestion
        if (error && error != ERROR_ALREADY_EXISTS) &&
```

(I'm not 100% sure whether we want to unmap in the success case or just any case besides ERROR_ALREADY_EXISTS. Trusting that you know.)

Reply:
> In fact, it definitely should, because `GetLastError()` _is not_ guaranteed to return 0 when the last call was successful - it only updates on errors.

For create calls that use open-always semantics, such as `CreateFile()` and `CreateFileMapping()`, on success the last error is set to either 0 or `ERROR_ALREADY_EXISTS`. In our case, we need to handle `ERROR_ALREADY_EXISTS` as an error, and close the opened handle.

Reply:
I was trying to be clever and avoid referring to `error` when it might not have been defined. But of course it's always defined (at least to zero if nothing's got in the way) so all you need is an `if (error != ERROR_ALREADY_EXISTS)`. Which I've now done.